database/dtb.ksy

```diff
@@ -0,0 +1,127 @@
+meta:
+  id: dtb
+  title: Flattened Devicetree Format
+  file-extension: dtb
+  application:
+    - Linux
+    - Das U-Boot
+  xref:
+    wikidata: Q16960371
+  license: CC0-1.0
+  encoding: ASCII
+  endian: be
+doc: |
+  Also referred to as Devicetree Blob (DTB). It is a flat
+  binary encoding of data (primarily devicetree data, although
+  other data is possible as well).
+
+  On Linux systems that support this the blobs can be accessed in
+  /sys/firmware/fdt :
+
+  - https://www.kernel.org/doc/Documentation/ABI/testing/sysfs-firmware-ofw
+
+  The encoding of strings used in the strings block and struct block is
+  actually a subset of ASCII:
+
+  https://github.com/devicetree-org/devicetree-specification/blob/master/source/chapter2-devicetree-basics.rst
+
+  Example files:
+
+  - https://github.com/qemu/qemu/tree/master/pc-bios
+doc-ref:
+  - https://github.com/devicetree-org/devicetree-specification/releases/tag/v0.3
+  - https://github.com/devicetree-org/devicetree-specification/blob/ba2aa67/source/flattened-format.rst
+  - https://elinux.org/images/f/f4/Elc2013_Fernandes.pdf
+seq:
+  - id: magic
+    -orig-id: magic
+    contents: [0xd0, 0x0d, 0xfe, 0xed]
+  - id: total_size
+    -orig-id: totalsize
+    type: u4
+  - id: ofs_structure_block
+    -orig-id: off_dt_struct
+    type: u4
+  - id: ofs_strings_block
+    -orig-id: off_dt_strings
+    type: u4
+  - id: ofs_memory_reservation_block
+    -orig-id: off_mem_rsvmap
+    type: u4
+  - id: version
+    -orig-id: version
+    type: u4
+  - id: last_compatible_version
+    -orig-id: last_comp_version
+    type: u4
+    valid:
+      max: version
+  - id: boot_cpuid_phys
+    -orig-id: boot_cpuid_phys
+    type: u4
+  - id: strings_block_size
+    -orig-id: size_dt_strings
+    type: u4
+  - id: structure_block_size
+    -orig-id: size_dt_struct
+    type: u4
+instances:
+  memory_reservation_block:
+    pos: ofs_memory_reservation_block
+    size: ofs_structure_block - ofs_memory_reservation_block
```

Comment:
According to https://github.com/devicetree-org/devicetree-specification/releases/download/v0.3/devicetree-specification-v0.3.pdf (section **5.3 Memory Reservation Block > 5.3.2 Format**, page 47/56), this lacks inner structure:

> ### 5.3.2 Format
> 
> The memory reservation block consists of a list of pairs of 64-bit big-endian integers, each pair being represented by the following C structure.
> 
> ```c
> struct fdt_reserve_entry {
>     uint64_t address;
>     uint64_t size;
> };
> ```
> Each pair gives the physical address and size in bytes of a reserved memory region. These given regions shall not overlap each other. The list of reserved blocks shall be terminated with an entry where both address and size are equal to 0. Note that the address and size values are always 64-bit. On 32-bit CPUs the upper 32-bits of the value are ignored.
> 
> Each uint64_t in the memory reservation block, and thus the memory reservation block as a whole, shall be located at an 8-byte aligned offset from the beginning of the devicetree blob (see section 5.6).

Reply:
That is correct. I skipped this part, as for my particular use case it is not interesting. Also, none of the test files I looked at actually had anything useful in there, except the terminator.

Reply:
I have added this now as well.